File: _old/SlimFacebook/app/src/main/java/it/rignanese/leo/slimfacebook/PictureActivity.java
package it.rignanese.leo.slimfacebook;

import android.app.Activity;
import android.app.DownloadManager;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.support.v4.app.ActivityCompat;
import android.support.v4.content.ContextCompat;
import android.view.Menu;
import android.view.MenuItem;
import android.view.WindowManager;
import android.webkit.WebSettings;
import android.widget.Toast;

import java.io.File;
import java.util.Date;

import it.rignanese.leo.slimfacebook.utility.MyAdvancedWebView;


/**
 * SlimSocial for Facebook is an Open Source app realized by Leonardo Rignanese <rignanese.leo@gmail.com>
 * GNU GENERAL PUBLIC LICENSE  Version 2, June 1991
 * GITHUB: https://github.com/rignaneseleo/SlimSocial-for-Facebook
 */

public class PictureActivity extends Activity implements MyAdvancedWebView.Listener {
    private MyAdvancedWebView webViewPicture;//the main webView where is shown facebook
    private SharedPreferences savedPreferences;//contains all the values of saved preferences
    private DownloadManager downloadManager;
    private long idDownloadedFile = -1;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_picture);

        savedPreferences = PreferenceManager.getDefaultSharedPreferences(this); // setup the sharedPreferences
        downloadManager = (DownloadManager) getSystemService(Context.DOWNLOAD_SERVICE);

        BroadcastReceiver receiver = downloadCompletedReceiver;
        registerReceiver(receiver, new IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE));

        SetupPictureWebView();

        String pictureUrl = getIntent().getStringExtra("URL");
        webViewPicture.loadUrl(pictureUrl);
    }

    BroadcastReceiver downloadCompletedReceiver = new BroadcastReceiver() {
        @Override
        public void onReceive(Context context, Intent intent) {
            String action = intent.getAction();
            if (DownloadManager.ACTION_DOWNLOAD_COMPLETE.equals(action)) {
                long downloadId = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, 0);
                DownloadManager.Query query = new DownloadManager.Query();

                query.setFilterById(idDownloadedFile);
                Cursor c = downloadManager.query(query);
                if (c.moveToFirst()) {
                    int columnIndex = c.getColumnIndex(DownloadManager.COLUMN_STATUS);
                    if (DownloadManager.STATUS_SUCCESSFUL == c.getInt(columnIndex)) {
                        //get the url
                        String pictureUri = c.getString(c.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI));

                        // Create the URI from the media
                        File media = new File(Uri.parse(pictureUri).getPath());
                        Uri uri = Uri.fromFile(media);

                        //to share
                        Intent shareIntent = new Intent(Intent.ACTION_SEND);
                        shareIntent.setType("image/jpeg");
                        shareIntent.putExtra(Intent.EXTRA_STREAM, uri);
                        startActivity(Intent.createChooser(shareIntent, getString(R.string.share)));

                        //reset
                        idDownloadedFile = -1;
                    }
                }
            }
        }
    };

    private void DownloadPicture(String url, boolean share) {
        //check permission
        if (Build.VERSION.SDK_INT >= 23 && ContextCompat.checkSelfPermission(this,
                android.Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_DENIED) {
            //ask permission
            Toast.makeText(getApplicationContext(), getString(R.string.acceptPermissionAndRetry),
                    Toast.LENGTH_LONG).show();
            int requestResult = 0;
            ActivityCompat.requestPermissions(this,
                    new String[]{android.Manifest.permission.WRITE_EXTERNAL_STORAGE}, requestResult
            );
        } else {
            //download photo
            DownloadManager.Request request = new DownloadManager.Request(Uri.parse(url));
            request.setTitle("SlimSocial Download");

            request.allowScanningByMediaScanner();
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED);

            Date date = new Date();
            long timeMilli = date.getTime();
            String filename = "SlimSocial_" + timeMilli + ".jpg";
            if (savedPreferences.getBoolean("pref_useSlimSocialSubfolderToDownloadedFiles", false)) {
                filename = "SlimSocial/"+filename;
            }

            request.setDestinationInExternalPublicDir(Environment.DIRECTORY_PICTURES, filename);

            if (share)
                request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN);

            // get download service and enqueue file
            long _idDownloadedFile = downloadManager.enqueue(request);

            if (share)
                idDownloadedFile = _idDownloadedFile;
            else
                Toast.makeText(getApplicationContext(), getString(R.string.downloadingPhoto),
                        Toast.LENGTH_LONG).show();
        }
    }

    //*********************** MENU ****************************
    //add my menu
    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.picture_menu, menu);
        return true;
    }

    //handling the tap on the menu's items
    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        int id = item.getItemId();
        switch (id) {
            case R.id.share: {//scroll on the top of the page
                DownloadPicture(webViewPicture.getUrl(), true);
                break;
            }
            case R.id.download: {//scroll on the top of the page
                DownloadPicture(webViewPicture.getUrl(), false);
                break;
            }
            default:
                break;
        }

        return super.onOptionsItemSelected(item);
    }

    //*********************** WEBVIEW ****************************
    private void SetupPictureWebView() {
        webViewPicture = findViewById(R.id.webViewPicture);
        webViewPicture.setListener(this, this);

        //  webViewPicture.setDesktopMode(true);

        //webViewPicture.requestFocus(View.F);
        getWindow().setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE);//remove the keyboard issue


        WebSettings settings = webViewPicture.getSettings();

        settings.setBuiltInZoomControls(true);
        settings.setUseWideViewPort(true);
        settings.setJavaScriptEnabled(true);
        settings.setLoadWithOverviewMode(true);
        settings.setDisplayZoomControls(false);
    }

    @Override
    public void onPageStarted(String url, Bitmap favicon) {
    }

    @Override
    public void onPageFinished(String url) {
    }

    @Override
    public void onPageError(int errorCode, String description, String failingUrl) {
    }

    @Override
    public void onDownloadRequested(String url, String suggestedFilename, String mimeType, long contentLength, String contentDisposition, String userAgent) {
    }

    @Override
    public void onExternalPageRequest(String url) {
    }

    @Override
    public boolean shouldLoadUrl(String url) {
        return true;
    }
}

File: SlimFacebook/app/src/main/java/it/rignanese/leo/slimfacebook/PictureActivity.kt
package it.rignanese.leo.slimfacebook

import android.Manifest
import android.app.Activity
import android.app.DownloadManager
import android.content.*
import android.content.pm.PackageManager
import android.graphics.Bitmap
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.os.Environment
import android.preference.PreferenceManager
import android.support.v4.app.ActivityCompat
import android.support.v4.content.ContextCompat
import android.view.Menu
import android.view.MenuItem
import android.view.WindowManager
import android.widget.Toast
import it.rignanese.leo.slimfacebook.utility.MyAdvancedWebView
import java.io.File

/**
 * SlimSocial for Facebook is an Open Source app realized by Leonardo Rignanese <rignanese.leo></rignanese.leo>@gmail.com>
 * GNU GENERAL PUBLIC LICENSE  Version 2, June 1991
 * GITHUB: https://github.com/rignaneseleo/SlimSocial-for-Facebook
 */
class PictureActivity : Activity(), MyAdvancedWebView.Listener{
    private lateinit var webViewPicture //the main webView where is shown facebook
            : MyAdvancedWebView
    private lateinit var savedPreferences //contains all the values of saved preferences
            : SharedPreferences
    private lateinit var downloadManager: DownloadManager
    private var idDownloadedFile: Long = -1
    override fun onCreate(savedInstanceState: Bundle) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_picture)
        savedPreferences = PreferenceManager.getDefaultSharedPreferences(this) // setup the sharedPreferences
        downloadManager = getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
        val receiver = downloadCompletedReceiver
        registerReceiver(receiver, IntentFilter(DownloadManager.ACTION_DOWNLOAD_COMPLETE))
        SetupPictureWebView()
        val pictureUrl = intent.getStringExtra("URL")
        webViewPicture!!.loadUrl(pictureUrl)
    }

    var downloadCompletedReceiver: BroadcastReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context, intent: Intent) {
            val action = intent.action
            if (DownloadManager.ACTION_DOWNLOAD_COMPLETE == action) {
                val downloadId = intent.getLongExtra(DownloadManager.EXTRA_DOWNLOAD_ID, 0)
                val query = DownloadManager.Query()
                query.setFilterById(idDownloadedFile)
                val c = downloadManager!!.query(query)
                if (c.moveToFirst()) {
                    val columnIndex = c.getColumnIndex(DownloadManager.COLUMN_STATUS)
                    if (DownloadManager.STATUS_SUCCESSFUL == c.getInt(columnIndex)) {
                        //get the url
                        val pictureUri = c.getString(c.getColumnIndex(DownloadManager.COLUMN_LOCAL_URI))

                        // Create the URI from the media
                        val media = File(Uri.parse(pictureUri).path)
                        val uri = Uri.fromFile(media)

                        //to share
                        val shareIntent = Intent(Intent.ACTION_SEND)
                        shareIntent.type = "image/jpeg"
                        shareIntent.putExtra(Intent.EXTRA_STREAM, uri)
                        startActivity(Intent.createChooser(shareIntent, getString(R.string.share)))

                        //reset
                        idDownloadedFile = -1
                    }
                }
            }
        }
    }

    private fun DownloadPicture(url: String, share: Boolean) {
        //check permission
        if (Build.VERSION.SDK_INT >= 23 && ContextCompat.checkSelfPermission(this,
                        Manifest.permission.WRITE_EXTERNAL_STORAGE) == PackageManager.PERMISSION_DENIED) {
            //ask permission
            Toast.makeText(applicationContext, getString(R.string.acceptPermissionAndRetry),
                    Toast.LENGTH_LONG).show()
            val requestResult = 0
            ActivityCompat.requestPermissions(this, arrayOf(Manifest.permission.WRITE_EXTERNAL_STORAGE), requestResult
            )
        } else {
            //download photo
            val request = DownloadManager.Request(Uri.parse(url))
            request.setTitle("SlimSocial Download")
            request.allowScanningByMediaScanner()
            request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
            var path = Environment.getExternalStoragePublicDirectory(Environment.DIRECTORY_PICTURES).path
            if (savedPreferences!!.getBoolean("pref_useSlimSocialSubfolderToDownloadedFiles", false)) {
                path += "/SlimSocial"
            }
            request.setDestinationInExternalPublicDir(path, "SlimSocial.jpg")
            if (share) request.setNotificationVisibility(DownloadManager.Request.VISIBILITY_HIDDEN)

            // get download service and enqueue file
            val _idDownloadedFile = downloadManager!!.enqueue(request)
            if (share) idDownloadedFile = _idDownloadedFile else Toast.makeText(applicationContext, getString(R.string.downloadingPhoto),
                    Toast.LENGTH_LONG).show()
        }
    }

    //*********************** MENU ****************************
    //add my menu
    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.picture_menu, menu)
        return true
    }

    //handling the tap on the menu's items
    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        val id = item.itemId
        when (id) {
            R.id.share -> {
                //scroll on the top of the page
                DownloadPicture(webViewPicture!!.url, true)
            }
            R.id.download -> {
                //scroll on the top of the page
                DownloadPicture(webViewPicture!!.url, false)
            }
            else -> {
            }
        }
        return super.onOptionsItemSelected(item)
    }

    //*********************** WEBVIEW ****************************
    private fun SetupPictureWebView() {
        webViewPicture = findViewById(R.id.webViewPicture)
        webViewPicture.setListener(this, this)

        //  webViewPicture.setDesktopMode(true);

        //webViewPicture.requestFocus(View.F);
        window.setSoftInputMode(WindowManager.LayoutParams.SOFT_INPUT_ADJUST_RESIZE) //remove the keyboard issue
        val settings = webViewPicture.settings
        settings.builtInZoomControls = true
        settings.useWideViewPort = true
        settings.javaScriptEnabled = true
        settings.loadWithOverviewMode = true
        settings.displayZoomControls = false
    }

    override fun onPageStarted(url: String?, favicon: Bitmap?) {

    }

    override fun onPageFinished(url: String?) {}
    override fun onPageError(errorCode: Int, description: String?, failingUrl: String?) {

    }

    override fun onDownloadRequested(url: String?, suggestedFilename: String?, mimeType: String?, contentLength: Long, contentDisposition: String?, userAgent: String?) {

    }
    override fun onExternalPageRequest(url: String?) {}
    override fun shouldLoadUrl(url: String?): Boolean {
        return true
    }
}